proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,649 @@
+     ``forall``-bound type variable, even though the ``forall`` binding is
+     closer to the occurrence.
+
+   * Built-in syntax for tuples and lists within types-in-terms is interpreted
+     as in terms. That is, ``[a]`` is a singleton list rather than the type of
+     a list, and ``(a, b)`` is a pair rather than the type of a pair.
+
+   * In type-checking, types-in-terms are rejected outright, unless converted
+     to a type by the "T2T-Mapping" (see point 8 below).
+
+5. Extend the name resolution rules: when looking up a term-level identifier
+   fails, look for a type-level identifier as fallback. As the result, a
+   term-level expression can also contain type variables and type constructors,
+   which will be either rejected or mapped to types (see "T2T-Mapping" below).
+
+6. In a term-level context, a reference to an ambiguous name (such as
+   ``Proxy``) is always resolved in favor of the data constructor, never the
+   type constructor.
+
+7. In a function application ``f e`` for some expressions ``f`` and ``e``, the
+   type of ``f`` affects neither parsing nor renaming of ``e``.
+
+8. In type checking, we alternate between two distinct modes: *checking* and
+   *inference*. This idea, called bidirectional type checking, is presented in
+   more detail in `"A quick look at impredicativity" <https://www.microsoft.com/en-us/research/uploads/prod/2020/01/quick-look-icfp20.pdf>`_.
+
+   * In inference mode, we never infer ``forall x -> t``.
+
+   * In checking mode, in a function application chain ``f e1 e2 e3``, we
+     follow the rules shown in Figure 4 of "A quick look impredicativity",
+     extended as follows::
+
+        rho = t2t(e)
+        G |- sigma[rho/a]; pis  ~>  Theta; phis; rho_r
+        ---------------------------------------------------------  ITVDQ
+        G |- (forall a -> sigma); e,pis  ~>  Theta; phis; rho_r
+
+     See "T2T-Mapping" below for an informal definition of ``t2t``.
+
+     In other words, given ``f :: forall a -> t``, the ``x`` in ``f x`` is
+     parsed and renamed as a term, but then mapped to a type.
+
+9.  Add a warning ``-Widentifier-forall``, included in ``-Wcompat``, which
+    warns on identifiers named ``forall``. In three releases, make ``forall``
+    a keyword everywhere, regardless of enabled extensions (same motivation as
+    `#193 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0193-forall-keyword.rst>`_).
+
+T2T-Mapping
+~~~~~~~~~~~
+
+T2T (term-to-type) mapping is a syntactic transformation, recursively applied
+to each subexpression. It occurs after name resolution, and thus can
+distinguish between term-level and type-level identifiers by examining the
+namespace.
+
+* A type constructor ``T`` is mapped to a type directly, without modification.
+
+* A type variable ``t`` is mapped to a type directly, without modification.
+
+* The types-in-terms (such as ``a -> b``, ``a => b``, ``forall a. b``) are
+  mapped to types directly, without modification.
```

Comment:
There are two ways to implement it.

1. The implementation can represent them as an additional constructor of `HsExpr`:

   ```
   data HsExpr pass
     = ...
     | TypeInTerms (HsType pass)
    ```

   So there will be no need for recursive modification:

   ```
   t2t (TypeInTerms t) = return t
   ```

2. The implementation can add separate constructors for the new syntactic forms:

   ```
   data HsExpr pass
     = ...
     | HsFunArrExp
                        (HsArrow pass)
                        (LHsExpr pass)   -- function type
                        (LHsExpr pass)
     | ...
    ```
    Then it would require recursive modification:
    ```
    t2t (HsFunArrExp arr e1 e2) = HsFunTy arr <$> t2t e1 <*> t2t e2
    ```

It hard to say which implementation plan is better, so I leave this sentence open to interpretation.

Reply:
But your specification says that it parses as `expr -> expr`, doesn't it? So, this part should at least be consistent with this choice.

Reply:
The grammar we use does not fully determine the AST we produce, as we have some overloading machinery. For example, expressions and patterns use one grammar, but produce either `Pat GhcPs` or `HsExpr GhcPs`. So even if the grammar is `expr -> expr`, we may be getting `HsType GhcPs` way before `t2t` enters the picture.